library/src-3.x/dotty/internal/StringContextMacro.scala

```diff
@@ -1,22 +1,758 @@
-      val parts2 = parts.head :: parts.tail.map(x => if (x.startsWith("%s")) x else "%s" + x)
-      parts2.mkString.format($args: _*)
+  /** This trait defines a tool to report errors/warnings that do not depend on Position. */
+  trait Reporter {
+
+    /** Reports error/warning of size 1 linked with a part of the StringContext.
+      *
+      * @param message the message to report as error/warning
+      * @param index the index of the part inside the list of parts of the StringContext
+      * @param offset the index in the part String where the error is
+      * @return an error/warning depending on the function
+      */
+    def partError(message : String, index : Int, offset : Int) : Unit
+    def partWarning(message : String, index : Int, offset : Int) : Unit
+
+    /** Reports error linked with an argument to format.
+      *
+      * @param message the message to report as error/warning
+      * @param index the index of the argument inside the list of arguments of the format function
+      * @return an error depending on the function
+      */
+    def argError(message : String, index : Int) : Unit
+
+    /** Reports error linked with the list of arguments or the StringContext.
+      *
+      * @param message the message to report in the error
+      * @return an error
+      */
+    def strCtxError(message : String) : Unit
+    def argsError(message : String) : Unit
+
+    /** Claims whether an error or a warning has been reported
+      *
+      * @return true if an error/warning has been reported, false
+      */
+    def hasReported() : Boolean
+
+    /** Stores the old value of the reported and reset it to false */
+    def resetReported() : Unit
+
+    /** Restores the value of the reported boolean that has been reset */
+    def restoreReported() : Unit
+  }
+
+  /** Retrieves a String from an Expr containing it
+   *
+   *  @param expression the Expr containing the String
+   *  @return the String contained in the given Expr
+   *  quotes an error if the given Expr does not contain a String
+   */
+  private def literalToString(expression : Expr[String])(implicit reflect: Reflection) : String = expression match {
+    case Const(string : String) => string
+    case _ => QuoteError("Expected statically known literal", expression)
+  }
+
+  /** Retrieves the parts from a StringContext, given inside an Expr, and returns them as a list of Expr of String
+   *
+   *  @param strCtxExpr the Expr containing the StringContext
+   *  @return a list of Expr containing Strings, each corresponding to one parts of the given StringContext
+   *  quotes an error if the given Expr does not correspond to a StringContext
+   */
+  def getPartsExprs(strCtxExpr : Expr[scala.StringContext])(implicit reflect : Reflection): List[Expr[String]] = {
+    import reflect._
+    strCtxExpr.unseal.underlyingArgument match {
+      case Apply(Select(Select(_, "StringContext") | Ident("StringContext"), "apply"), List(parts1)) =>
```

Comment:
This should be implementable with 
```scala
case ('{ new StringContext(${ExprSeq(parts)}: _*) }, ExprSeq(args)) =>
```
as in the `f-interpolator-neg`. Or something similar.

Reply:
This was modified on commit 1e6ce2a. Thanks for pointing this out.